hugr-passes/src/static_circuit.rs

```diff
@@ -0,0 +1,398 @@
+            extensions: self.extensions.clone(),
+            alloc_ops: self.alloc_ops.clone(),
+            free_ops: self.free_ops.clone(),
+        }
+    }
+}
+
+impl<'a, H: HugrView> StaticCircuitContext<'a, H> {
+    fn assert_invariants(&self) {
+        debug_assert!(self.alloc_ops.is_disjoint(&self.free_ops));
+        for &n in self.alloc_ops.iter().chain(&self.free_ops) {
+            debug_assert!(self.hugr.valid_node(n));
+            let Some(e) = self.hugr.get_optype(n).as_extension_op() else {
+                panic!("node {n} is not in extensions")
+            };
+            debug_assert!(self.extensions.contains(e.def().extension()));
+        }
+    }
+}
+
+impl<'a, H: HugrView> ConstLoader<QbHistory> for StaticCircuitContext<'a, H> {}
+
+impl<'a, H> Deref for StaticCircuitContext<'a, H> {
+    type Target = H;
+
+    fn deref(&self) -> &Self::Target {
+        self.hugr
+    }
+}
+
+impl<'a, H: HugrView> DFContext<QbHistory> for StaticCircuitContext<'a, H> {
+    type View = H;
+
+    fn interpret_leaf_op(
+        &self,
+        node: Node,
+        e: &ExtensionOp,
+        ins: &[PartialValue<QbHistory>],
+        outs: &mut [PartialValue<QbHistory>],
+    ) {
+        self.assert_invariants();
+        let extension = e.def().extension();
+        if !self.extensions.contains(extension) {
+            return;
+        }
+
+        if ins.iter().any(|x| x == &PartialValue::Bottom) {
+            return;
+        }
+
+        // assert_eq!(ins.len(), self.hugr.num_inputs(node));
+        // assert_eq!(outs.len(), self.hugr.num_outputs(node));
+
+        let qb_ins = qubits_in(self.hugr, node).collect::<Vec<_>>();
+        let qb_outs = qubits_out(self.hugr, node).collect::<Vec<_>>();
+
+        if qb_ins.is_empty() && qb_outs.is_empty() {
+            return;
+        }
+
+        if let Some(&qb_i) = self.alloc_ops.get(&node) {
+            assert_eq!(0, qb_ins.len());
+            assert_eq!(1, qb_outs.len());
+            for (i, out) in outs.iter_mut().enumerate() {
+                if qb_outs[0] == i.into() {
+                    out.join_mut(QbHistory::init(qb_i).into());
+                } else {
+                    out.join_mut(PartialValue::Top);
+                }
+            }
+        } else if self.free_ops.contains(&node) {
+            assert_eq!(0, qb_outs.len());
+            for out in outs.iter_mut() {
+                out.join_mut(PartialValue::Top);
+            }
+        } else {
+            assert_eq!(qb_outs.len(), qb_ins.len(), "{e:?}");
+            let mb_in_qbs = qb_ins
+                .iter()
+                .map(|&i| match &ins[i.index()] {
+                    PartialValue::Value(qbh) => Some((i, qbh)),
+                    _ => None,
+                })
+                .collect::<Option<HashMap<_, _>>>();
+            let gating_qbs = mb_in_qbs
+                .iter()
+                .flat_map(|hash_map| hash_map.iter().map(|(_, qbh)| qbh.qb_index()))
```

Comment:
So the "gating_nodes" of a new node is the ordered list of input gates, right? That could go in a comment on `Gate::gating_nodes`.

I'm *slightly* concerned that `Node` (rather than `(Node, Port)`) is sufficiently accurate - the wire coming from one or another port sounds like an important difference - I think the counterargument is that if a different port were used, *because of linearity* other places would see different `Node` instead, so we might be OK.